File: src/ipc/quota.ts
import { invoke } from "@tauri-apps/api/core";

export interface QuotaItem {
  name: string;
  pct: number;
  used: number | null;
  total: number | null;
  remaining: number | null;
  resets_in: string | null;
  resets_at_ms: number | null;
  /** Currency unit for balance-based items (e.g. "CNY", "USD"). Null for
   *  window-based quota. */
  unit: string | null;
  /** True for monetary-balance items (OpenRouter credits, Moonshot balance):
   *  no reset-window semantics, and keep-alive never pings them. */
  is_balance: boolean;
}

export interface QuotaExtractorFields {
  /** JSON path (dot-separated, numeric segments index arrays). */
  name?: string;
  used?: string;
  remaining?: string;
  total?: string;
  unit?: string;
}

/// User-configured balance extractor: GET a URL and pull fields from the
/// JSON response by dot-path. `{{baseUrl}}` / `{{apiKey}}` in url/headers
/// are substituted from the endpoint. Custom extracts are always
/// balance-shaped (is_balance): no reset window, keep-alive never pings.
export interface QuotaExtractorConfig {
  enabled: boolean;
  url: string;
  headers?: Record<string, string>;
  /** Static currency unit used when `fields.unit` is not configured. */
  unit?: string | null;
  fields: QuotaExtractorFields;
}

/// A built-in provider quota fetcher — the dispatch key for the `preset`
/// query plan. Mirrors `endpoint_quota::BuiltinKind` (snake_case on the wire).
export type BuiltinKind = "zai" | "minimax" | "openrouter" | "opencode_go" | "mock";

/// How an endpoint's quota is queried — the unified "query plan". Mirrors
/// `endpoint_quota::QuotaQueryPlan` (internally tagged on `source`). `none`
/// means no query configured (quota display + keep-alive stay gated until
/// the user picks one). Plans are declared by provider presets and stamped
/// at create time; legacy endpoints resolve a plan from their base_url host.
export type QuotaQueryPlan =
  | { source: "none" }
  | { source: "preset"; kind: BuiltinKind }
  | ({ source: "custom" } & QuotaExtractorConfig);

export interface RefreshEndpointConfig {
  enabled: boolean;
  protocol: string | null;
  model: string | null;
  target_quota_name: string | null;
  last_status: string | null;
  check_rate_secs: number;
  reset_grace_secs: number;
  /** Custom quota extractor; when enabled it overrides the built-in fetch.
   *  Legacy — kept for compat with older blobs; new writes use query_plan. */
  extractor: QuotaExtractorConfig | null;
  /** Explicit query plan (the canonical "how is quota queried" choice).
   *  `null` means "use legacy resolution" (extractor.enabled or host detect). */
  query_plan?: QuotaQueryPlan | null;
  /** True once any fetch has returned data for the current plan. Gates both
   *  the keep-alive switch and the quota bars. Cleared when the plan changes. */
  provisioned?: boolean | null;
  /** Quota windows shown on the provider-card preview (multi-select).
   *  `null` (legacy) falls back to the 5h-name heuristic; `[]` = show none.
   *  Display-only — independent of `target_quota_name` (the keep-alive ping
   *  target). Set from the Quota page's settings dialog. */
  preview_windows?: string[] | null;
  /** OpenCode Go dashboard workspace ID (non-secret). Paired with the `auth`
   *  cookie stored encrypted under `opencode-go-cookie-{id}`; edited in the
   *  Quota settings dialog's creds section. Must be carried through every
   *  full-blob settings write (`quota_refresh_set_settings` replaces the
   *  whole blob) or the server's serde default silently erases it. */
  opencode_workspace_id?: string | null;
}

export interface RefreshSettings {
  endpoints: Record<string, RefreshEndpointConfig>;
}
export interface EndpointQuota {
  ok: boolean;
  plan: string | null;
  error: string | null;
  items: QuotaItem[];
}
export const endpointFetchQuota = (id: string) =>
  invoke<EndpointQuota>("endpoint_fetch_quota", { id });

// ---- 5h quota auto-refresh ----
export const quotaRefreshGet = () =>
  invoke<RefreshSettings>("quota_refresh_get_settings");
export const quotaRefreshSet = (value: RefreshSettings) =>
  invoke<void>("quota_refresh_set_settings", { value });

/// OpenCode Go dashboard credentials status. The cookie value is never
/// returned — only whether one is set (credential boundary, like API keys).
export interface OpencodeCredsStatus {
  workspace_id: string | null;
  has_cookie: boolean;
}
export const opencodeGetCreds = (endpointId: string) =>
  invoke<OpencodeCredsStatus>("opencode_get_creds", { endpointId });
export const opencodeSetCreds = (
  endpointId: string,
  cookie: string,
  workspaceId: string,
) =>
  invoke<void>("opencode_set_creds", {
    endpointId,
    cookie,
    workspaceId,
  });

export interface PingPreview {
  method: string;
  url: string;
  headers: [string, string][];
  body: string;
  protocol: string;
  model: string;
}
export const quotaKeepalivePreview = (endpointId: string) =>
  invoke<PingPreview>("quota_keepalive_preview", { endpointId });
export interface PingNowResult {
  ok: boolean;
  status: string;
}
export const quotaPingNow = (endpointId: string) =>
  invoke<PingNowResult>("quota_ping_now", { endpointId });

export type KeepAlivePhase =
  | "disabled"
  | "not_configured"
  | "unverified"
  | "idle"
  | "resetting"
  | "pinging"
  | "retrying"
  | "error";
export interface KeepAliveState {
  phase: KeepAlivePhase;
  last_success_at: number | null;
  next_fire_at: number | null;
  last_error: string | null;
  attempts: number;
  /// Epoch ms of the most recent worker heartbeat (one per tick), shared
  /// across all endpoints since the worker is a single thread. `null` before
  /// the first tick completes. A stale value means the worker is dead or the
  /// process is suspended.
  last_heartbeat_at: number | null;
  /// Epoch ms of the last worker panic recovered by the supervisor, or
  /// `null` if none (or it has since ticked successfully).
  last_panic_at: number | null;
}
export const quotaKeepaliveStatus = (endpointId: string) =>
  invoke<KeepAliveState>("quota_keepalive_status", { endpointId });

File: src-tauri/src/commands/quota.rs
use crate::error::{AppError, AppResult};
use crate::{db, secrets};
use serde::Serialize;
use super::run_blocking;
use tauri::State;

// ---- 5h-quota auto refresh settings ----

#[tauri::command]
pub fn quota_refresh_get_settings(
    state: State<'_, crate::AppState>,
) -> AppResult<crate::quota_refresh::RefreshSettings> {
    let conn = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
    crate::quota_refresh::load_settings(&conn)
}

#[tauri::command]
pub fn quota_refresh_set_settings(
    state: State<'_, crate::AppState>,
    value: crate::quota_refresh::RefreshSettings,
) -> AppResult<()> {
    let conn = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
    crate::quota_refresh::save_settings(&conn, &value)
}

/// OpenCode Go dashboard credentials — the workspace ID (non-secret) and the
/// `auth` session cookie (secret). Returned to the UI WITHOUT the cookie
/// value: only whether one is set, mirroring how API keys are never surfaced.
#[derive(Debug, Clone, Serialize)]
pub struct OpencodeCredsStatus {
    pub workspace_id: Option<String>,
    pub has_cookie: bool,
}

#[tauri::command]
pub fn opencode_get_creds(
    state: State<'_, crate::AppState>,
    endpoint_id: String,
) -> AppResult<OpencodeCredsStatus> {
    let conn = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
    let settings = crate::quota_refresh::load_settings(&conn)?;
    let workspace_id = settings
        .endpoints
        .get(&endpoint_id)
        .and_then(|c| c.opencode_workspace_id.clone());
    let has_cookie = secrets::get(&format!("opencode-go-cookie-{endpoint_id}"))?
        .is_some_and(|c| !c.is_empty());
    Ok(OpencodeCredsStatus { workspace_id, has_cookie })
}

/// Store the OpenCode Go dashboard credentials. The cookie is encrypted at
/// rest (`secrets.rs`, AES-256-GCM, key `opencode-go-cookie-{id}`); the
/// workspace ID is non-secret and lives in the settings blob. Changing the
/// credentials clears the verified state so a fetch must re-confirm data
/// before the bars/keep-alive unlock.
#[tauri::command]
pub fn opencode_set_creds(
    state: State<'_, crate::AppState>,
    endpoint_id: String,
    cookie: String,
    workspace_id: String,
) -> AppResult<()> {
    let cookie = cookie.trim();
    if !cookie.is_empty() {
        secrets::set(&format!("opencode-go-cookie-{endpoint_id}"), cookie)?;
    }
    let conn = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
    crate::quota_refresh::update_settings(&conn, |settings| {
        crate::quota_refresh::set_opencode_workspace_id(settings, &endpoint_id, &workspace_id);
    })
}

#[tauri::command]
pub fn quota_keepalive_preview(
    state: State<'_, crate::AppState>,
    endpoint_id: String,
) -> AppResult<crate::quota_refresh::PingPreview> {
    let conn = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
    let ep = db::get_endpoint(&conn, &endpoint_id)?
        .ok_or_else(|| AppError::Validation(format!("unknown endpoint '{endpoint_id}'")))?;
    let settings = crate::quota_refresh::load_settings(&conn)?;
    let cfg = settings.endpoints.get(&endpoint_id).cloned().unwrap_or_default();
    crate::quota_refresh::build_ping_preview(&ep, &cfg)
}

/// Result of an on-demand keep-alive ping fired by the popover `[test]`
/// link. On failure, `status` carries the full body snippet (no
/// truncation) so the user can see exactly what the server said.
#[derive(Debug, Clone, Serialize)]
pub struct PingNowResult {
    pub ok: bool,
    pub status: String,
}

#[tauri::command]
pub async fn quota_ping_now(
    state: State<'_, crate::AppState>,
    endpoint_id: String,
) -> AppResult<PingNowResult> {
    let db = state.db.clone();
    run_blocking(move || {
        let (ep, cfg) = {
            let conn = db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
            let ep = db::get_endpoint(&conn, &endpoint_id)?
                .ok_or_else(|| AppError::Validation(format!("unknown endpoint '{endpoint_id}'")))?;
            let settings = crate::quota_refresh::load_settings(&conn)?;
            let cfg = settings
                .endpoints
                .get(&endpoint_id)
                .cloned()
                .unwrap_or_default();
            (ep, cfg)
        };
        let key = match secrets::get(&endpoint_id) {
            Ok(Some(k)) if !k.is_empty() => k,
            _ => {
                return Ok(PingNowResult {
                    ok: false,
                    status: "no API key configured".into(),
                })
            }
        };
        crate::quota_refresh::update_keepalive(&endpoint_id, |s| {
            s.phase = crate::quota_refresh::KeepAlivePhase::Pinging
        });
        let res = crate::quota_refresh::try_ping(&ep, &cfg, &key);
        crate::quota_refresh::record_ping_outcome(&endpoint_id, &res);
        let status = match &res {
            Ok(()) => "ok".to_string(),
            Err(f) if f.transient => format!("retrying: {}", f.message),
            Err(f) => format!("error: {}", f.message),
        };
        let conn = db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
        let _ = crate::quota_refresh::set_status_public(&conn, &endpoint_id, &status);
        Ok(PingNowResult {
            ok: res.is_ok(),
            status,
        })
    })
    .await
}

/// Runtime keep-alive state for one endpoint (phase, last success, next
/// fire, error, attempts). Read-only and lock-free — in-memory only, so it
/// needs no DB connection.
#[tauri::command]
pub fn quota_keepalive_status(
    endpoint_id: String,
) -> crate::quota_refresh::KeepAliveState {
    crate::quota_refresh::keepalive_state(&endpoint_id)
}
